core/lib/makeswift/utils/use-products.ts

```diff
@@ -40,14 +41,20 @@ export function useProducts({ collection, collectionLimit = 20, additionalProduc
     fetcher,
   );
 
-  const combinedProducts = [
-    ...(collectionData?.products.slice(0, collectionLimit) ?? []),
-    ...(additionalData?.products ?? []),
-  ];
+  const combinedProducts = useMemo(
+    () => [
+      ...(collectionData?.products.slice(0, collectionLimit) ?? []),
+      ...(additionalData?.products ?? []),
+    ],
+    [collectionData, additionalData, collectionLimit],
+  );
 
   const isLoading = isCollectionLoading || isAdditionalLoading;
 
-  const products = isLoading ? null : combinedProducts.map(bcProductToVibesProduct);
+  const products = useMemo(
+    () => (isLoading ? null : combinedProducts.map(bcProductToVibesProduct)),
+    [isLoading, combinedProducts, bcProductToVibesProduct],
+  );
```

Comment:
Essentially, we were returning a new array of "combined" products every time, triggering downstream re-renders.

Reply:
Do we know why downstream re-renders were resulting in the loading state being displayed? I would've thought that only happens when the component re-mounts. Were re-renders resulting in re-mounts?

Reply:
> Do we know why downstream re-renders were resulting in the loading state being displayed? I would've thought that only happens when the component re-mounts. Were re-renders resulting in re-mounts?

Yes, `Streamable.all` was constructing a new promise for each new value. Either of these two fixes (making the product list stable or eliminating an unnecessary promise) eliminates the flicker. I chose to fix both.

Reply:
Ah, makes sense. I would hope the pending state showing was a result of a new promise being passed in such that the stable values aren't necessary. Often adding `useMemo` and `useCallback` adds more overhead than it saves in terms of work on re-renders. Thanks for clarifying 🙏🏻